File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int minValue(TreeNode root){
        TreeNode temp = root;
        int mini = 0;
        while(temp != null){
            mini = temp.val;
            temp = temp.left;
        }
        return mini;
    }
    public TreeNode deleteNode(TreeNode root, int key) {
        if(root == null)
            return root;
        
        if(root.val == key){
            if(root.left == null && root.right == null){
                root = null;
                return root;
            }
            
            else if(root.left == null && root.right != null){
                TreeNode temp = root.right;
                root = null;
                return temp;
            }
            
            else if(root.left != null && root.right == null){
                TreeNode temp = root.left;
                root = null;
                return temp;
            }
            
            // if(root.left != null && root.right != null){
            else{
                int mini = minValue(root.right);
                root.val = mini;
                root.right = deleteNode(root.right, mini);
                return root;
            }
        }
        
        else if(root.val > key){
            root.left = deleteNode(root.left, key);
            return root;
        }
        
        else{
            root.right = deleteNode(root.right, key);
            return root;
        } 
    }
}

File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.cpp
class Solution {
public:
    TreeNode* deleteNode(TreeNode* root, int key) {
        // Base case: the key was not found or the tree is empty
        if (!root) return nullptr;

        // 1. Search for the node
        if (key < root->val) {
            root->left = deleteNode(root->left, key);
        } 
        else if (key > root->val) {
            root->right = deleteNode(root->right, key);
        } 
        // 2. Node found! Let's delete it.
        else {
            // Case 1 & 2: Node has 0 or 1 child
            if (!root->left) {
                TreeNode* rightChild = root->right;
                delete root;
                return rightChild;
            } 
            else if (!root->right) {
                TreeNode* leftChild = root->left;
                delete root;
                return leftChild;
            }
            
            // Case 3: Node has 2 children
            // Find the inorder successor (smallest node in the right subtree)
            TreeNode* successor = root->right;
            while (successor->left) {
                successor = successor->left;
            }
            
            // Replace the current node's value with the successor's value
            root->val = successor->val;
            
            // Delete the successor node from the right subtree
            root->right = deleteNode(root->right, successor->val);
        }
        
        return root;
    }
};